File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
from collections import deque

def solution(info, edges):
    answer = 0
    n = len(info)
    tree = {i: [] for i in range(n)}
    for p, s in edges: tree[p].append(s)
    q = deque([[0, tree[0], 1, 0]])
    
    while q:
        now, can_move, sheep, wolf = q.popleft()
        if answer < sheep: answer = sheep
        for i, node in enumerate(can_move):
            if info[node] == 1:
                if sheep > wolf + 1:
                    q.append([node, can_move[:i] + can_move[i + 1:] + tree[node], sheep, wolf + 1])
            else:
                q.append([node, can_move[:i] + can_move[i + 1:] + tree[node], sheep + 1, wolf])

    return answer

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.js
function solution(info, edges) {
    let answer = 0;
    const q = [[0, 1, 0, new Set()]];

    function buildTree(info, edges) {
        const tree = Array.from({ length: info.length }, () => []);
        for (let edge of edges) tree[edge[0]].push(edge[1]);
        return tree;
    }
    const tree = buildTree(info, edges);

    while (q.length) {
        const [cur, sheep, wolf, visited] = q.shift();
        answer = Math.max(answer, sheep);
        const newVisited = new Set(visited);
        tree[cur].forEach(node => newVisited.add(node));
        
        newVisited.forEach(nNode => {
            if (info[nNode]) {
                if (sheep !== wolf + 1) {
                    const nextVisited = new Set(newVisited);
                    nextVisited.delete(nNode);
                    q.push([nNode, sheep, wolf + 1, nextVisited]);
                }
            } else {
                const nextVisited = new Set(newVisited);
                nextVisited.delete(nNode);
                q.push([nNode, sheep + 1, wolf, nextVisited]);
            }
        });
    }

    return answer;
}
